File: src/main/java/co/uk/isxander/evergreenhud/config/ElementConfig.java
/*
 * Copyright (C) isXander [2019 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.config;

import co.uk.isxander.evergreenhud.elements.Element;
import co.uk.isxander.evergreenhud.elements.ElementManager;
import co.uk.isxander.xanderlib.utils.Constants;
import co.uk.isxander.xanderlib.utils.json.BetterJsonObject;
import co.uk.isxander.evergreenhud.EvergreenHUD;
import com.google.gson.JsonArray;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ElementConfig implements Constants {

    public static final int VERSION = 3;
    public static final File OLD_CONFIG_FOLDER = new File(EvergreenHUD.DATA_DIR, "elements");
    public static final File CONFIG_FILE = new File(EvergreenHUD.DATA_DIR, "elements.json");

    private final ElementManager manager;

    public ElementConfig(ElementManager manager) {
        this.manager = manager;
    }

    public BetterJsonObject generateJson() {
        BetterJsonObject root = new BetterJsonObject();
        root.addProperty("version", VERSION);

        JsonArray array = new JsonArray();
        for (Element element : manager.getCurrentElements()) {
            BetterJsonObject obj = new BetterJsonObject();
            obj.addProperty("type", element.getType());
            obj.add("settings", element.generateJson());
            array.add(obj.getData());
        }
        root.getData().add("elements", array);

        return root;
    }

    public void save() {
        BetterJsonObject root = generateJson();
        root.writeToFile(CONFIG_FILE);
    }

    public void load() {
        if (!CONFIG_FILE.exists() && OLD_CONFIG_FOLDER.exists()) {
            loadOld();

            save();
            return;
        }

        BetterJsonObject root;
        try {
            root = BetterJsonObject.getFromFile(CONFIG_FILE);
        } catch (Exception e) {
            save();
            return;
        }

        boolean v2 = false;
        if (root.optInt("version") == 2) {
            EvergreenHUD.LOGGER.warn("Converting Element Config v2 -> v3. Conversion issues may arise");
            v2 = true;
        } else if (root.optInt("version") != VERSION) {
            EvergreenHUD.LOGGER.warn("Resetting configuration! Older or newer config version detected.");
            EvergreenHUD.getInstance().notifyConfigReset();
            save();
            return;
        }

        JsonArray array = root.getData().getAsJsonArray("elements");
        boolean finalVer = v2;
        array.forEach((jsonElement) -> {
            BetterJsonObject elementConfig = new BetterJsonObject(jsonElement.getAsJsonObject());
            String id = elementConfig.optString("type");
            Element element = manager.getNewElementInstance(id);
            if (element != null) {
                BetterJsonObject settings = elementConfig.getObj("settings");
                if (finalVer) {
                    element.loadJsonOld(settings);
                } else {
                    element.loadJson(settings);
                }

                manager.addElement(element);
            } else {
                EvergreenHUD.LOGGER.warn("Found unknown element id: " + id + ". Skipped.");
            }
        });

        if (v2) save();
    }

    public void loadOld() {
        EvergreenHUD.LOGGER.info("Converting old configuration system.");

        List<Element> availableElements = new ArrayList<>();
        manager.getAvailableElements().forEach((name, clazz) ->
            availableElements.add(manager.getNewElementInstance(name))
        );

        for (File config : Objects.requireNonNull(OLD_CONFIG_FOLDER.listFiles())) {
            for (Element element : availableElements) {
                if (config.getName().equalsIgnoreCase(element.getMetadata().getName() + ".json")) {
                    BetterJsonObject root;
                    try {
                        root = BetterJsonObject.getFromFile(config);
                    } catch (Exception e) {
                        save();
                        return;
                    }

                    if (root.optBoolean("enabled")) {
                        element.loadJsonOld(root);

                        manager.addElement(element);
                    }

                    break;
                }
            }
            config.delete();
        }
        OLD_CONFIG_FOLDER.delete();
    }

}

File: core/src/main/kotlin/dev/isxander/evergreenhud/config/ElementConfig.kt
/*
 * EvergreenHUD - A mod to improve on your heads-up-display.
 * Copyright (C) isXander [2019 - 2021]
 *
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-2.1.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/isXander/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package dev.isxander.evergreenhud.config

import com.electronwill.nightconfig.core.Config
import com.electronwill.nightconfig.core.file.FileConfig
import com.electronwill.nightconfig.core.io.WritingMode
import dev.isxander.evergreenhud.EvergreenHUD
import dev.isxander.evergreenhud.api.logger
import dev.isxander.evergreenhud.elements.Element
import dev.isxander.evergreenhud.elements.ElementManager
import dev.isxander.evergreenhud.utils.jsonFormat
import dev.isxander.evergreenhud.utils.jsonWriter
import java.io.File

class ElementConfig(private val manager: ElementManager) {
    private var shouldSave = false

    fun save() {
        val data = Config.of(jsonFormat)
        data.set<Int>("schema", SCHEMA)

        val arr = mutableListOf<Config>()
        for (element in manager) {
            arr.add(Config.of(jsonFormat).apply {
                set<String>("id", element.metadata.id)
                set<Config>("data", element.conf)
            })
        }
        data.set<List<Config>>("elements", arr)

        CONFIG_FILE.parentFile.mkdirs()
        jsonWriter.write(data, CONFIG_FILE, WritingMode.REPLACE)
        shouldSave = false
    }

    fun load() {
        if (!CONFIG_FILE.exists()) save().also { return@load }
        val data = attemptConversion(FileConfig.of(CONFIG_FILE).apply { load() })

        val arr: List<Config> = data["elements"]
        for (elementData in arr) {
            val id = elementData["id"] ?: "null"
            val element = manager.getNewElementInstance<Element>(id)

            if (element == null) {
                logger.err("Found unknown element id ($id) in json! This probably means someone tampered with the json!")
                continue
            }

            element.conf = elementData["data"]
            manager.addElement(element)
        }

        if (shouldSave) save()
    }

    @Suppress("UNUSED_EXPRESSION")
    private fun attemptConversion(data: Config): Config {
        val currentSchema = data["schema"] ?: 0

        // corrupt config. Reset
        if (currentSchema == 0 || currentSchema > SCHEMA) {
            return Config.of(jsonFormat)
        }

        // there is no point recoding every conversion
        // when a new schema comes to be
        // so just convert the old conversions until done
        var convertedData = data
        var convertedSchema = currentSchema
        while (convertedSchema != SCHEMA) {
            logger.info("Converting element configuration v$convertedSchema -> v${convertedSchema + 1}")
            when (convertedSchema) {

            }
            convertedSchema++
        }

        return convertedData
    }

    companion object {
        const val SCHEMA = 4
        val CONFIG_FILE: File
            get() = File(EvergreenHUD.profileManager.profileDirectory, "elements.json")
    }
}
